ReactAndroid/src/main/java/com/facebook/react/uimanager/UIManagerModule.java

```diff
@@ -83,6 +83,14 @@ public UIManagerModule(
     reactContext.addLifecycleEventListener(this);
   }
 
+  /**
+   * This method gives an access to the underlying native view hierarchy manager. Note that all the
+   * operations executed on that object should be run in the UI thread. Use with caution
+   */
+  public NativeViewHierarchyManager getNativeViewHierarchyManager() {
+    return mUIImplementation.getUIViewOperationQueue().getNativeViewHierarchyManager();
```

Comment:
Not happy with that approach either. I used to have a different way of setting up `NativeAnimatedModule` but decided to go with this instead. Here are some of the aspects driving the decision:
 1. I like the idea of using native module abstraction to expose API for managing animated nodes, starting/stopping animation etc. In my first prototype I put all those methods to UIManager. As it is already overloaded with a number of different methods (even `scrollTo` used to be a method of `UIManager`) I believe it's better for the animated native API to be extracted away from the UIManager
 2. Since animated API is extracted as a separate module (see (1)) I though it would be a good idea to make it "optional" so that developer can decide whether to include it on the RN "package" list (note that this module will be bringing "rebound" library as a dependency)

My "old" approach was to instantiate NativeAnimatedModule inside of UIManager and pass NativeHierarchyManager as a constructor argument to it. See corresponding changes in  [`UIManagerModule.java`](https://github.com/kmagiera/react-native/blob/1e9333d668a1959b180403cf3cd6267c4f01e67a/ReactAndroid/src/main/java/com/facebook/react/uimanager/UIManagerModule.java#L89). The issue with this approach was that it would require for the animated module to be added to the `CoreModulesPackage` (see it [here](https://github.com/kmagiera/react-native/blob/1e9333d668a1959b180403cf3cd6267c4f01e67a/ReactAndroid/src/main/java/com/facebook/react/CoreModulesPackage.java#L87)) and therefore would prevent us from making that module "optional" (If you think it should be a "core" module I'm fine with that too, just didn't want to make the decision for you).

Last thing re alternative implementations of `UIImplementation` - I didn't forget about it, just didn't know its still a thing. However, in order to drive animations from UI thread we need to have a synchronous way of updating the UI while `UIImplementation` is asynchronous (at least the default implementation of it). Don't have any good ideas on how can this be solved. One thing that comes to my mind is to add method to `UIImplementation` - `synchronouslyUpdateViewOnUIThread` as that's the only method we'll ever need to access from the perspective of native animated module.